### ISSUE TITLE
startup-timeout not happening on all worker processes

### ISSUE THREAD

**Issue by coderanger** [GrahamDumpleton/mod_wsgi#369]
So I have this daemon process group:

```
WSGIDaemonProcess prod-summon-platform-web1 processes=4 threads=37 display-name=%{GROUP} python-home=/home/ubuntu/platform_env python-path=/home/ubuntu/summon-platform startup-timeout=6
```

But during a deploy where we expect Django to lock up, 2 of the 4 worker procs are not doing the startup-timeout self-kill. The other two stuck around until I manually killed them, despite mod_wsgi definitely seeing that the WSGI script couldn't be loaded:

```
[Wed Nov 07 14:30:08.964537 2018] [wsgi:error] [pid 11962:tid 139974631716608] [remote ...:15064] mod_wsgi (pid=11962): Target WSGI script '/home/ubuntu/summon-platform/summon_platform/wsgi.py' cannot be loaded as Python module.
[Wed Nov 07 14:30:08.964559 2018] [wsgi:error] [pid 11962:tid 139974631716608] [remote ...:15064] mod_wsgi (pid=11962): Exception occurred processing WSGI script '/home/ubuntu/summon-platform/summon_platform/wsgi.py'.
```

On top of that I see some really weird time issues in the logs, that I'm guessing is related. The Apache error log is zigzagging between 14:27 and 20:27 (the deploy happened at 20:27 UTC). Is there any chance that setting a non-UTC timezone in Django can confuse things? That sets `os.environ['TZ'] = self.TIME_ZONE` somewhere deep in Django.

**Comment by coderanger** [comment]
Confirmed that Django app is set to `TIME_ZONE: America/Chicago`. So that would explain the 6 hour shift I guess, if Apache passively accepts the zone set inside Django. Would that break startup-timeout somehow?

**Comment by GrahamDumpleton** [comment]
What version of mod_wsgi are you using?

I screwed up the implementation at one point. So it was not working as intended for versions 4.5.10 through 4.5.15. Some distro still ship out of date version.

From 4.5.16 release notes.

```
* The feature behind the ``startup-timeout`` option of ``WSGIDaemonProcess``
  was broken by prior fix related to feature in 4.5.10. This meant the option
  was not resulting in daemon processes being restarted when the WSGI script
  file could not be loaded successfully by the specified timeout.
```

Seems the change in 4.5.10 was to fix another issue with it.

```
* Startup timeout wasn't being cancelled after succesful load of the WSGI
  script file and instead was only being done after first request had
  finished. This meant that if first request took longer than the startup
  timeout the process would be wrongly restarted.
```

So you want to make sure 4.5.16 or later is used.

**Comment by GrahamDumpleton** [comment]
BTW, a high number of threads such as ``threads=37`` is always dangerous. Should only do this if absolutely sure your application is very highly I/O bound and waiting most of the time. The sweet spot is usually at most 3-5 threads per process and scale up number of processes instead of threads.

**Comment by coderanger** [comment]
4.5.17, custom build package, so it should be okay on the implementation. I'm trying to figure out from https://github.com/GrahamDumpleton/mod_wsgi/blob/3781411da928e66c3ade6d00ca836b422e8551eb/src/server/mod_wsgi.c#L9234 what would happen if the ambient timezone changed during loading. I would think `apr_time_now` should be in epoch seconds so timezone should have no effect on it, right? (and I know the threads are dumb, trying to keep the current Ansible deploy on life support while I get the k8s-y replacement ready :)

**Comment by GrahamDumpleton** [comment]
Time zone only affects the log messages.

Are you using ``WSGIImportScript``, or using ``WSGIScriptAlias`` with both ``process-group`` and ``application-group`` arguments set on it at the same time, as opposed to an exception. These trigger pre-loading. Wondering whether may an issue with that.

**Comment by coderanger** [comment]
```
  WSGIPassAuthorization On
  WSGIProcessGroup prod-summon-platform-web1
  WSGIScriptAlias / /home/ubuntu/summon-platform/summon_platform/wsgi.py
```

**Comment by coderanger** [comment]
I can also pull some of the `loglevel info`'d logs for you if that would help :-)

**Comment by GrahamDumpleton** [comment]
Yeah, wouldn't hurt to see ``info`` level logs to see whether timeout did get logged but didn't shutdown, or whether timeout never even occurred.

**Comment by coderanger** [comment]
I've hopefully redacted all the user data in this: https://gist.github.com/coderanger/e35664cf8a7b29749e3fdcaf155ef6a0

**Comment by coderanger** [comment]
Line 630 is where I manually SIGTERM'd the two rogue worker procs (PID 11962 and 11963). You can see above that, they were still spewing the Django grossness :)

**Comment by GrahamDumpleton** [comment]
Reviewing old issues that haven't been closed out.

Of note in respect of ``startup-timeout`` is that the timeout period isn't started until the first request is handled by the process. So, if WSGI script preloading is used via ``WSGIImportScript`` or by specifying both ``process-group`` and ``application-group`` to ``WSGIScriptAlias``, you will see a failure when the script is preloaded, but unless requests are sent to Apache and the daemon process receives a request, the daemon process will just sit there and will not reload. If there is sufficient requests coming in, they should get distributed evenly and the process will get a request, at which point the timeout period will start.

That said, in the above configuration there was no preloading.

**Comment by coderanger** [comment]
No longer at that job so I can't access the config anymore but requests would have some in pretty constantly from Kubernetes readiness probes.

**Comment by GrahamDumpleton** [comment]
Okay. Wasn't expecting a response and this is just me recording information as I think through what can check, or extra logging can add to help track down issues in future. Thanks for letting me know though.

**Comment by GrahamDumpleton** [comment]
I actually think I may have found the problem. Doesn't occur with ``mod_wsgi-express``, only a manually configured Apache/mod_wsgi. Specifically, if ``startup-timeout`` and ``restart-interval`` were both specified, with no other timeouts such as ``graceful-timeout`` or ``request-timeout``, the ``restart-interval`` timeout was overriding the ``startup-timeout`` for when it should next check to see if the process should be shutdown for some reason. End result is that if you had ``restart-interval`` set to 3600 (1 hour), then that is what the startup timeout effectively ended up being.

This is the second time today where I have uncovered the reason for an old issue because in the past when I had first looked at it, had only tested and evaluated the problem in the context of how ``mod_wsgi-express`` worked and not considered that raw mod_wsgi doesn't specify defaults for some timeouts. :-(

**Comment by GrahamDumpleton** [comment]
Turned out to be a bit more complicated than first thought.

At worst case the delay in process restart would have been the lessor of the time periods specified by the options ``restart-interval``, ``deadlock-timeout``, ``graceful-timeout`` and ``eviction-timeout``. If ``request-timeout`` were defined it would however still be calculated correctly. As ``request-timeout`` was by default defined when using ``mod_wsgi-express``, this issue only usually affected mod_wsgi when manually configuring Apache.

Since ``deadlock-timeout`` defaults to 300 seconds, that would likely have been how long it took to restart, even though ``restart-interval`` may have been longer.

Anyway, fixed now for mod_wsgi 4.9.0, so going to close this issue now.
